### PR TITLE
Refactor the C++ guidelines

### LINKED ISSUE #2114: C++ Guidelines: Fit content to new Guidelines structure
## Description

Per the MQ Guidelines initiative to fill gaps in the guidelines across languages, we are moving to a new common structure for the guidelines and calling out areas where there are content gaps -- i.e. one or more of the other Tier 1 language guidelines have guidance that is not language-specific that is not represented in this language's guidelines.

**Note**

For Implementation, the below structure is based on .NET content.  It is intended to provide implementation details for things that aren't visible in the public API, with a TOC that largely maps to the API Design section in the Design TOC, with sections filled for areas where you already have content written.  Feel free to reach out with any questions.

## New Structure
(Numbers for ease of reading here, but won't be shown in the guidelines page TOC)

For further details, please see the restructured .NET guidelines [here](https://guidelinescollab.github.io/azure-sdk/dotnet_introduction.html).

### Design
1. Introduction
    1. Design Principles
    2. General Guidelines  
    3. Non-HTTP Libraries
2. Azure SDK API Design
    1. Service Client
        1. Service Client Constructors
            1. Client Configuration
            2. Setting the Service Version
            3. Sync and Async
            4. Client Immutability
        2. Service Methods
            1. Naming
            2. Return Types
            3. Cancellation
        4. Service Method Parameters
            1. Options Parameters
            2. Parameter Validation
        5. Methods Returning Collections (Paging)
        6. Methods Invoking Long Running Operations
        7. Conditional Request Methods	
        8. Hierarchical Clients
    2. Supporting Types
        1. Model Types
        2. Enumerations
        3. Using Azure Core Types
        4. Using Primitive Types
    3. Exceptions (Errors)
    4. Authentication
    5. Namespaces
    6. Support for Mocking

3. Azure SDK Libraries (Packages) 
    1. Packaging
        1. Service-Specific Common Libraries
    2. Versioning
    3. Dependencies
    4. Native Code
    5. Doc Comments
4. Repository Guidelines
    1. Documentation Style
    2. README
    3. Samples

### Implementation
1. API Implementation
    1. Service Client
        1. Service Methods
            1. HttpPipeline
            2. HttpPipelinePolicy/Custom Policies
        2. Service Method Parameters
            1.  Parameter Validation
    2. Supporting Types
            1. Serialization
            2. Enumeration-like Structs
2. SDK Feature Implementation
    1. Configuration
    2. Logging
    3. Distributed Tracing	
    4. Telemetry
3. Testing 
4. Language-specific other

### PR DESCRIPTION
**Pull request by RickWinter**
Refactor the C++ guidelines to align with the new common structure.

**NOTICE** This change is not adding/removing sections.  
It is step 1  which moves all existing content into the most relevant sections of the new common structure.  
Step 2 will update/add/remove content to fill in the missing gaps.

Closes #2114

### CODE REVIEW

File: docs/cpp/introduction.md [hunk 1]
```diff
@@ -8,19 +8,21 @@ sidebar: general_sidebar
...
 
 * The SDK should follow the general design guidelines and conventions for the target language. It should feel natural to a developer in the target language.
 * We embrace the ecosystem with its strengths and its flaws.
 * We work with the ecosystem to improve it for all developers.
 
-### Consistent
+**Consistent**
 
 * Client libraries should be consistent within the language, consistent with the service and consistent between all target languages. In cases of conflict, consistency within the language is the highest priority and consistency between all target languages is the lowest priority.
 * Service-agnostic concepts such as logging, HTTP communication, and error handling should be consistent. The developer should not have to relearn service-agnostic concepts as they move between client libraries.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:28]: "must be consistent"
"tHE DEVELOPER MUST NOT HAVE TO RELEASRN..."

File: docs/cpp/introduction.md [hunk 2]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+|:-------------------------|------------------|
+| gcc                      | `-Wall -Wextra`  |
+| cpp and XCode            | `-Wall -Wextra`  |
+| MSVC                     | `/W4`            |
+
+When configuring your client library, particular care must be taken to ensure that the consumer of your client library can properly configure the connectivity to your Azure service both globally (along with other client libraries the consumer is using) and specifically with your client library.
+
+
+### The Service Client {#cpp-client}
+
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:127]: Who will write these 2 sections?

File: docs/cpp/introduction.md [hunk 3]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+1. Log level, which must take effect immediately across the Azure SDK.
+2. Tracing on/off, which must take effect immediately across the Azure SDK.
+
+{% include requirement/MUSTNOT id="cpp-config-noruntime" %} use client library specific runtime
+configuration such as environment variables or a config file. Keep in mind that many IOT devices
+won't have a filesystem or an "environment block" to read from.
+
+##### Using ClientOptions {#cpp-usage-options}
+
+> TODO: This section needs to be driven by code in the Core library.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:161]: Who is writing these?

File: docs/cpp/introduction.md [hunk 4]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-versioning-select-api-version" %} allow the consumer to explicitly select a supported service API version when instantiating the service client.
+
+Use a constructor parameter called `version` on the client options type.
+
+* The `version` parameter must be the first parameter to all constructor overloads.
+* The `version` parameter must be required, and default to the latest supported service version.
+* The type of the `version` parameter must be `ServiceVersion`; an enum nested in the options type.
+* The `ServiceVersion` enum must use explicit values starting from 1.
+* `ServiceVersion` enum value 0 is reserved. When 0 is passed into APIs, ArgumentException should be thrown.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:177]: Do we have an ArgumentExpcetion?
Is there a reason why we can't use 0? Seems like it would be OK to me.

File: docs/cpp/introduction.md [hunk 5]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+> TODO: This section needs to be driven by code in the Core library.
+
+#### Service Methods {#cpp-client-methods}
+
+> TODO: This section needs to be driven by code in the Core library.
+
+_Service methods_ are the methods on the client that invoke operations on the service.
+
+##### Sync and Async
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:189]: You could kill this section as we're sync only and for the foreseeable future.

File: docs/cpp/introduction.md [hunk 6]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-design-naming-variables-local" %} name local variables and parameters with **camelCase**.
+
+{% highlight cpp %}
+// Examples of the above naming rules:
+
+namespace Azure { namespace Group { namespace Service {
+int PublicNamespaceScopeVariable; // these should be used sparingly
+const int c_PublicNamespaceScopeConstant = 42;
+constexpr int c_OtherPublicNamespaceScopeConstant = 42;
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:234]: Can we reccommend const or constexpr; not both?

File: docs/cpp/introduction.md [hunk 7]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-design-naming-variables-public-global" %} name namespace scope non-constant variables intended only for internal consumption with a `g_` prefix followed by **camelCase**. For example, `g_applicationContext`. Note that all such cases will be in a `Details` namespace or an unnamed namespace.
+
+{% include requirement/MUST id="cpp-design-naming-variables-local" %} name local variables and parameters with **camelCase**.
+
+{% highlight cpp %}
+// Examples of the above naming rules:
+
+namespace Azure { namespace Group { namespace Service {
+int PublicNamespaceScopeVariable; // these should be used sparingly
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:232]: Change "thsese" to "public fields shoud be used..."

File: docs/cpp/introduction.md (hunk 6 above)
Comment (JeffreyRichter) [reply docs/cpp/introduction.md:234]: Also, I thought we killed off c_ prefix for constants? Am I wrong?

File: docs/cpp/introduction.md [hunk 8]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+std::chrono::milliseconds Timeout;
+uint32 MyWeightKg;
+{% endhighlight %}
+
+{% include requirement/MUST id="cpp-design-naming-variables-one-per-line" %} Declare or define each variable on its own line, except when declaring bitfields. An exception can be made when declaring bitfields (to clarify that the variable is a part of one bitfield). The use of bitfields in general is discouraged.
+
+
+##### Cancellation
+
+{% include requirement/MUST id="cpp-service-methods-cancellation" %} ensure all service methods, both asynchronous and synchronous, take an optional `Context` parameter called _context_.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:274]: We onyl have synchronous and Context is a field in an option struct now.

File: docs/cpp/introduction.md [hunk 9]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-params-client-validation" %} validate client parameters.
+
+{% include requirement/MUSTNOT id="cpp-params-server-validation" %} validate service parameters.  This includes null checks, empty strings, and other common validating conditions. Let the service validate any request parameters.
+
+{% include requirement/MUST id="cpp-params-check-devex" %} validate the developer experience when the service parameters are invalid to ensure appropriate error messages are generated by the service.  If the developer experience is compromised due to service-side error messages, work with the service team to correct prior to release.
+
+#### Methods Returning Collections (Paging) {#cpp-paging}
+
+Although object-orientated languages can eschew low-level pagination APIs in favor of high-level abstractions, C acts as a lower level language and thus embraces pagination APIs provided by the service.  You should work within the confines of the paging system provided by the service.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:348]: "C++ acts..."
Although we may change this is we come up with a paging pattern.

File: docs/cpp/introduction.md [hunk 10]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+##### Model Type Naming
+
+> TODO: This section needs to be driven by code in the Core library.
+
+#### Enumerations
+
+> TODO: This section needs to be driven by code in the Core library.
+
+{% include requirement/MUST id="cpp-enums" %} use an `enum` for parameters, properties, and return types when values are known.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:397]: "...when values are known and there is certainty that the set of enum symbols will NEVER change over all versions of the service."

File: docs/cpp/introduction.md [hunk 11]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+The `Azure::Core::ETag` type is located in `azure-core` package.
+
+{% include requirement/MUST id="cpp-primitives-uri" %} use `Azure::Core::Uri` to represent URIs.
+
+The `Azure::Core::Uri` type is located in `azure-core` package.
+
+### Exceptions {#cpp-errors}
+
+Error handling is an important aspect of implementing a client library. It is the primary method by which problems are communicated to the consumer. Because we intend for the C client libraries to be used on a wide range of devices with a wide range of reliability requirements, it's important to provide robust error handling.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:421]: "C++ client libraries..."

File: docs/cpp/introduction.md [hunk 12]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+        // missing throw;
+    }
+}
+{% endhighlight %}
+
+### Authentication {#cpp-authentication}
+
+Azure services use a variety of different authentication schemes to allow clients to access the service. Conceptually, there are two entities responsible in this process: a credential and an authentication policy.  Credentials provide confidential authentication data.  Authentication policies use the data provided by a credential to authenticate requests to the service.
+
+{% include requirement/MUST id="cpp-design-logical-client-support-all-auth-techniques" %} support all authentication techniques that the service supports and are available to a client application (as opposed to service side).  C is used only for client applications when talking to Azure, so some authentication techniques may not be valid.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:533]: Search for "C " and change to "C++ " throughout.

File: docs/cpp/introduction.md [hunk 13]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-design-logical-client-prefer-token-auth" %} provide credential types that can be used to fetch all data needed to authenticate a request to the service in a non-blocking atomic manner for each authentication scheme that does not have an implementation in Azure Core.
+
+{% include requirement/MUST id="cpp-apisurface-auth-in-constructors" %} provide service client constructors or factories that accept any supported authentication credentials.
+
+{% include requirement/SHOULDNOT id="cpp-design-logical-client-surface-no-connection-strings" %} support providing credential data via a connection string. Connection string interfaces should be provided __ONLY IF__ the service provides a connection string to users via the portal or other tooling.
+
+{% include requirement/MUSTNOT id="cpp-design-logical-client-surface-no-connection-string-ctors" %} support constructing a service client with a connection string unless such connection string. Provide a `CreateFromConnectionString` static member function which returns a client instead to encourage customers to choose non-connection-string-based authentication.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:543]: bad grammar  on 1st sentence.

File: docs/cpp/introduction.md [hunk 14]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+If your service implements a non-standard credential system (one that is not supported by Azure Core), then you need to produce an authentication policy for the HTTP pipeline that can authenticate requests given the alternative credential types provided by the client library.
+
+{% include requirement/MUST id="cpp-implementing-secure-auth-erase" %} Use a "secure" function to zero authentication or authorization credentials as soon as possible once they are no longer needed. Examples of such functions
+include: `SecureZeroMemory`, `memset_s`, and `explicit_bzero`. Examples of insecure functions include `memset`. An optimizer may notice that the credentials are
+never accessed again, and optimize away the call to `memset`, resulting in the credentials remaining in memory.
+
+{% include requirement/MUST id="cpp-implementing-auth-policy" %}
+provide a suitable authentication policy that authenticates the HTTP request in the HTTP pipeline when using non-standard credentials.  This includes custom connection strings, if supported.
```
Comment (JeffreyRichter) [inline docs/cpp/introduction.md:557]: I'm not sure what this is saying.
The auth policy applies the auth to the request; it doesn't "authenticate teh HTTP request".
Connection strings should be converted to a credential by the static factory function and then the connection string is never used again. So, it's odd to talk about connections strings  and the pipeline so close ot each other. Connection strings are gone long before the pipeline enters the picture.

**Review by JeffreyRichter — APPROVED** [review]

File: docs/cpp/introduction.md (hunk 2 above)
Comment (RickWinter) [reply docs/cpp/introduction.md:127]: in the second pass I'll be working to flesh out and improve all the sections.  This first pass gets the topics in alignment with the structure.

File: docs/cpp/introduction.md (hunk 4 above)
Comment (RickWinter) [reply docs/cpp/introduction.md:177]: Agreed, I removed teh exception and the 0.  This snuck in from the template, good catch.

File: docs/cpp/introduction.md (hunk 6 above)
Comment (RickWinter) [reply docs/cpp/introduction.md:234]: Must have been missed.  I'll fix that up.

File: docs/cpp/introduction.md [hunk 15]
```diff
@@ -8,19 +8,21 @@ sidebar: general_sidebar
 
 {% include draft.html content="The C++ Language guidelines are in DRAFT status" %}
 
+## Introduction
```
Comment (ahsonkhan) [inline docs/cpp/introduction.md:11]: Update the `title` of this document to be `C++ Azure SDK Design Guidelines"` rather than `C++ Guidelines: Introduction`, and rename the file `design.md` rather than `introduction.md`.

File: docs/cpp/introduction.md [hunk 16]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+* Recoverable Error
+    : Things like trying to open a file that doesn't exist, or trying to write to a full disk. These kinds of errors can usually be handled by a function's caller directly, and need to be considered by callers that want to be robust.
+
+**Exhaustion / Act of God**
+
+{% include requirement/MUSTNOT id="cpp-design-logical-errorhandling-actofgod-no-return" %} return an error to the caller.
+
+{% include requirement/MUST id="cpp-design-logical-errorhandling-actofgod-crash" %} crash, if possible. This means calling some form of fast failing function, like `abort`.
+
+Note: if your client library needs to be resilient to these kinds of errors you must either provide a fallback system, or construct your code in a way to facilitate proving that such errors can not occur.
```
Comment (annelo-msft) [inline docs/cpp/introduction.md:442]: Is there an example of where this has been done in a Track 2 library?  Ideally, the guidelines would drive consistency here.

File: docs/cpp/introduction.md [hunk 17]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+        c();
+    } catch (...) {
+        // missing throw;
+    }
+}
+{% endhighlight %}
+
+### Authentication {#cpp-authentication}
+
+Azure services use a variety of different authentication schemes to allow clients to access the service. Conceptually, there are two entities responsible in this process: a credential and an authentication policy.  Credentials provide confidential authentication data.  Authentication policies use the data provided by a credential to authenticate requests to the service.
```
Comment (annelo-msft) [inline docs/cpp/introduction.md:531]: This section is very general.  It would be good to discuss this in the context of what C++ APIs look like here.

File: docs/cpp/introduction.md [hunk 18]
```diff
@@ -46,17 +48,1019 @@ The Azure SDK should be designed to enhance the productivity of developers conne
...
+
+{% include requirement/MUST id="cpp-docs-readme-consumer" %} optimize the `README.md` for the consumer of the client library.
+
+The contributor guide (`CONTRIBUTING.md`) should be a separate file linked to from the main component `README.md`.
+
+### Samples {#cpp-samples}
+
+> TODO: Provide sample guidelines.
+
+## Commonly Overlooked C++ API Design Guidelines {#cpp-appendix-overlookedguidelines}
```
Comment (annelo-msft) [inline docs/cpp/introduction.md:1057]: From what source are these guidelines overlooked?  In .NET, KC is calling out important guidelines from the .NET Framework Design Guidelines that ppl sometimes forget to follow.  It's great to call these out for C++ as well, but we should be clear what the source of truth is.

File: docs/cpp/implementation.md [hunk 19]
```diff
@@ -165,28 +170,25 @@ In general, our advice to consumers of these libraries is to establish logging i
...
+We believe testing is a part of the development process, so we expect unit and integration tests to be a part of the source code.  All components must be covered by automated testing, and developers should strive to test corner cases and main flows for every use case.
+
+All code should contain, at least, requirements, unit tests, end-to-end tests, and samples.
 
-{% include requirement/MUSTNOT id="cpp-dependencies-concrete" %} depend on concrete logging, dependency injection, or configuration technologies (except as implemented in the Azure Core library).  The client library will be used in applications that might be using the logging, DI, and configuration technologies of their choice.
+Tests should be written using the [Google Test][] library.
 
-### C and C++ language specifics
+### Language-specific other
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:191]: Let's call this C++ Language-Specific Guidance

File: docs/cpp/implementation.md [hunk 20]
```diff
@@ -8,114 +8,119 @@ sidebar: general_sidebar
...
 
-## Parameter validation
+- The value is retrieved and deserialized from service, and may contain a value not supported by the client library.
+- The value is roundtripped: the value is retrieved and deserialized from the service, and may later be serialized and sent back to the server.
 
-The service client will have several methods that perform requests on the service.  _Service parameters_ are directly passed across the wire to an Azure service.  _Client parameters_ are not passed directly to the service, but used within the client library to fulfill the request.  Examples of client parameters include values that are used to construct a URI, or a file that needs to be uploaded to storage.
+> TODO: Content in this section may need a new home.
 
-{% include requirement/MUST id="cpp-params-client-validation" %} validate client parameters.
+{% include requirement/MUST id="cpp-design-naming-enum" %} name `enum class`es and enumerators using **PascalCase**.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:70]: This feels like guidance regarding enums in the public-facing API.  Consider moving this into the API Design section?

File: docs/cpp/implementation.md [hunk 21]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
 
 {% include requirement/MUST id="cpp-dependencies-adparch" %} consult the [Architecture Board] if you wish to use a dependency that is not on the list of approved dependencies.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:199]: Should this and other dependency-related guidance go in the Dependencies section on the API Design page?

File: docs/cpp/implementation.md [hunk 22]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
 
-## Testing
+{% include requirement/MUSTNOT id="cpp-test-implicit-assign" %} use implicit assignment inside a test.  This is generally an accidental omission of the second `=` of the logical compare. The following is confusing and prone to error.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:201]: Should these go in the Testing subsection?

File: docs/cpp/implementation.md [hunk 23]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
 
-Tests should be written using the [Google Test][] library.
+{% highlight cpp %}
+a = b;
+if (a) { ... }
+{% endhighlight %}
+
+{% include requirement/MUSTNOT id="cpp-no-register" %} use the register keyword.  Modern compilers will put variables in registers automatically.
+
+{% include requirement/MUST id="cpp-be-const-correct" %} be `const` correct.  C++ provides the `const` keyword to allow passing as parameters objects that cannot change to indicate when a method doesn't modify its object. Using `const` in all the right places is called "const correctness."
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:216]: Is this relevant to the public API?

File: docs/cpp/implementation.md [hunk 24]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
-All code should contain, at least, requirements, unit tests, end-to-end tests, and samples.
+Does the programmer really mean assignment here? Sometimes yes, but usually no. Instead use explicit tests and avoid assignment with an implicit test. The recommended form is to do the assignment before doing the test:
 
-Tests should be written using the [Google Test][] library.
+{% highlight cpp %}
+a = b;
+if (a) { ... }
+{% endhighlight %}
+
+{% include requirement/MUSTNOT id="cpp-no-register" %} use the register keyword.  Modern compilers will put variables in registers automatically.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:214]: This should go in a section with a header.

File: docs/cpp/implementation.md [hunk 25]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+
+{% include requirement/MUST id="cpp-design-logical-rule-of-zero" %} In class types, implement the "rule of zero", the "rule of 3", or the "rule of 5". That is, of the special member functions, a type should implement exactly one of the following:
+
+* No copy constructor, no copy assignment operator, no move constructor, no move assignment operator, or destructor.
+* A copy constructor, a copy assignment operator, no move constructor, no move assignment operator, and a destructor.
+* A copy constructor, a copy assignment operator, a move constructor, a move assignment operator, and a destructor.
+
+This encourages use of resource managing types like std::unique_ptr (which implements the rule of 5) as a compositional tool in more complex data models that implement the rule of zero.
+
+{% include requirement/MUST id="cpp-design-logical-initialize-all-data" %} provide types which are usable when default-initialized. (That is, every constructor must initialize all type invariants, not assume members have default values of 0 or similar.)
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:394]: This feels like something that should be in the API design section on type design.

File: docs/cpp/implementation.md [hunk 26]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+    TypeWithInvariants a{}; // value-initializes a TypeWithInvariants, OK
+    TypeWithInvariants b; // default-initializes a TypeWithInvariants, we want this to be OK
+    BadTypeWithInvariants c{}; // value-initializes a BadTypeWithInvariants, OK
+    BadTypeWithInvariants d; // default-initializes a BadTypeWithInvariants, this will trigger
+                             // undefined behavior if anyone calls d.Next()
+}
+{% endhighlight %}
+
+
+{% include requirement/MUST id="cpp-design-logical-no-getters-or-setters" %} define getters and setters for data transfer objects.  Expose the members directly to users unless you need to enforce some constraints on the data.  For example:
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:425]: Public API?

File: docs/cpp/implementation.md [hunk 27]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+    void SetRetryTimeout(int i) noexcept {
+        if (i < 0) i = 0; // clamp i to the range [0, 1000]
+        if (i > 1000) i = 1000;
+        m_retryTimeout = i;
+    }
+};
+
+{% endhighlight %}
+
+{% include requirement/MUST id="cpp-design-logical-optimize-position" %} declare variables in structures organized by use in a manner that minimizes memory wastage because of compiler alignment issues and size.  All things being equal, use alphabetical ordering.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:481]: Public API?

File: docs/cpp/implementation.md [hunk 28]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+// Good
+struct Foo {
+    int A; // C is now stored in that padding
+    int C;
+    char *B;
+    char *D;
+};
+{% endhighlight %}
+
+{% include requirement/MUSTNOT id="cpp-design-logical-enumsareinternal" %} use enums to model any data sent to the service. Use enums only for types completely internal to the client library. For example, an enum to disable Nagle's algorithm would be OK, but an enum to ask the service to create a particular entity kind is not.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:501]: This should go in the public API sections.

File: docs/cpp/implementation.md [hunk 29]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+
+#### Const and Reference members
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:503]: Public API?

File: docs/cpp/implementation.md [hunk 30]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+        return m_maxRetryCount;
+    }
+};
+{% endhighlight %}
+
+#### Integer sizes
+
+The following integer rules are listed in rough priority order. Integer size selections are primarily driven by service future compatibility. For example, just because today a service might have a 2 GiB file size limit does not mean that it will have such a limit forever. We believe 64 bit length limits will be sufficient for sizes an individual program works with for the foreseeable future.
+
+{% include requirement/MUST id="cpp-design-logical-integer-files" %} Represent file sizes with `int64_t`, even on 32 bit platforms.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:535]: Also relevant to the public API?

File: docs/cpp/implementation.md [hunk 31]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
 
-{% include requirement/MUST id="cpp-style-filenames" %} use characters in the range `[a-z0-9_]` for the name portion (before the file extension).  No other characters are permitted.
+> TODO: Verify with the security team, and what are the alternatives?
 
-Filenames should be concise, but convey what role the file plays within the library.
+#### Enumerations
 
-{% include requirement/MUST id="cpp-style-headerguards" %} use `#pragma once`
+{% include requirement/MUST id="cpp-design-logical-client-enumerations-no-enums" %} use `enum` or `enum class` for values shared "over the wire" with a service, to support future compatibility with the service where additional values are added. Such values should be persisted as strings in client data structures instead.
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:553]: Public API.

Also, does this contradict the guidance above about not using enums in the public API?

File: docs/cpp/implementation.md [hunk 32]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+{% highlight cpp %}
+namespace {
+struct HashComputation {
+    int InternalBookkeeping;
+};
+} // unnamed namespace
 {% endhighlight %}
 
-{% include requirement/MAY id="cpp-style-whole-sdk-header" %} have a header file that includes an entire client library. For example, `<azure/speech.hpp>`.
+#### Class Types (including `union`s and `struct`s)
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:619]: Public API?

File: docs/cpp/implementation.md [hunk 33]
```diff
@@ -196,43 +198,469 @@ Unlike many other programming languages, which have large runtimes, the C++ stan
...
+// Bad: Uses C-style typedef:
+typedef struct IotClient {
+    char* ApiVersion;
+    IotClientCredentials* Credentials;
+    int RetryTimeout;
+} AzIotClient;
+{% endhighlight %}
 
-## Tooling
+#### Tooling
```
Comment (annelo-msft) [inline docs/cpp/implementation.md:664]: Below, where you discuss API Documentation ... this should live in the Documentation Comments section on the API Design page.

**Review by annelo-msft — COMMENTED** [review]
Please let me know if you have any questions!  It might make sense to set up some time to chat 1-1 about these -- happy to do that.

File: docs/cpp/implementation.md (hunk 19 above)
Comment (annelo-msft) [reply docs/cpp/implementation.md:191]: I thought we had already agreed that we won't have identical headers across languages?
